[quant][graphmode] Dynamic Quant : Do not depend on input shapes

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39458 [quant][graphmode] Dynamic quant: Insert observers for module output
* **#39412 [quant][graphmode] Dynamic Quant : Do not depend on input shapes**

Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

supriyar committed:
[quant][graphmode] Dynamic Quant : Do not depend on input shapes
Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -745,6 +866,55 @@ void propagateQuantizationOps(Block* block) {
+    auto& v = graph->inputs()[idx];
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      auto observer_name = findObserverName(v);
+      if (observer_name && isWeight(module, v)) {
```
Can you add a comment explaining how isWeight works?

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -745,6 +866,55 @@ void propagateQuantizationOps(Block* block) {
+      }
+    }
+  }
+  std::stack<Block*> blocks_to_visit;
```
For understanding, at this point, we have a stack of weight tensors in weight_values. What does the code below do? It seems to be finding other weights? Can you add a comment clarifying the logic?

supriyar replied:
It traverses all the blocks in the graph to get the observed weight values.

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2244,10 +2240,9 @@ def __init__(self):
         qconfig_dict = {'': default_dynamic_qconfig}
         model = torch.jit.script(M()).eval()
-        model = quantize_dynamic_script(model, qconfig_dict, data)
+        model = quantize_dynamic_script(model, qconfig_dict)
```
Is there a test for shared weights?
 For example, if I have a module of the form:
```python
class DynamicModel(torch.nn.Module):
    def __init__(self):
        super(DynamicModel, self).__init__()
        self.weight  = torch.rand(10,10)
        self.mod1 = myMod(self.weight)

 def forward(self, x):
       y = self.mod1(x)
       z = torch.nn.functional.linear(y, self.weight)
       return z

class myMod(torch.nn.Module):
    def __init__(wt):
        super(myMod, self).__init__()
        self.linear = nn.Linear(10, 10)
        self.linear.weight = wt

 def forward(self, x):
       return self.linear(x)
      
```
Can we handle cases where only myMod is quantized (has a qconfig) and the reverse (only MyMod is in float?)

raghuramank100 reviewed (commented):
This is great!. Some questions on test coverage

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -745,6 +870,61 @@ void propagateQuantizationOps(Block* block) {
+      // Ensure that we only track values that are the output of observer node.
+      // isWeight returns true if the observed value is a weight of quantized
+      // op.
+      if (observer_name && isWeight(module, v)) {
```
wait, how would input of the graph the output of the observer module?

supriyar replied:
Can remove this

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -745,6 +870,61 @@ void propagateQuantizationOps(Block* block) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end;) {
+      Node* n = *it++;
```
you can just use for each loop here I think, since there is no modification to the nodes

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -745,6 +870,61 @@ void propagateQuantizationOps(Block* block) {
+          weight_values.push_back(v);
+        }
+      }
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
```
can you check if this works for weight observers used in if?
e.g.
```python
def forward(self, x):
       if cond:
           x = F.linear(x, ...)
       else:
           x = F.linear(x, ...)
       return x
```

supriyar replied:
It works for nn.Linear. I am looking into making it work for F.linear as well.

jerryzh168 replied:
the weight for nn.Linear is in the method graph, and there is no `if` in that graph.

supriyar replied:
I think the main issue with `if` in the graph is caching the graph in `graph_to_subgraph_nodes_`. We call the extract code once for each weight value with the same graph input. Since this only provides a small optimization benefit I'm planning to remove this and make sure we correctly identify the weight subgraph even if the graph has multiple if statements. @jerryzh168 let me know if you think otherwise.

jerryzh168 replied:
I don't feel we need to remove the caching to support prim::If, why is that a problem?

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +529,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    Value* self,
+    Node* observer) {
+  Value* observed_weight = observer->output();
+  TORCH_CHECK(
```
should be TORCH_INTERNAL_ASSERT

jerryzh168 replied:
I think it's OK to remove this check, we are only calling this function with weight values anyways

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -430,6 +442,12 @@ class InsertQuantDeQuantHelper {
+
+  // Map from values in original graph to corresponding values in the extracted
+  // subgraph.
+  std::unordered_map<Value*, Value*> remap_values_in_observer_subgraph_;
```
can you separate the code that clones a graph to another helper class?

supriyar replied:
Is the main reason to move it this map? I can keep this local to `extractAndRunWeightObserver` and pass it in each time I build the subgraph.

jerryzh168 replied:
that's one concern yes, separate this to a different class also helps with the readability and maintainability

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +529,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    // Track values in the subgraph for which producer needs to be found.
+    std::unordered_set<Value*> subgraph_node_inputs;
+    updateCurrInputSet(observer->inputs().vec(), subgraph_node_inputs);
+    while (n && n != self->node()) {
```
I think it's better to check for input values against self value

jerryzh168 replied:
we should also check for https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit13hitGraphInputEPNS0_5ValueE
you can move this function to Value class I think

jerryzh168 replied:
since we need to reject the following case:
```python
def forward(self, x, m):
       weight = m(x)
       return F.linear(x, weight)
```

supriyar replied:
Not sure I follow, why would the case for node check vs input value check differ for the example above?

jerryzh168 replied:
value `m` in the example will have the same node as self, but it should be rejected, since in the end we are only going to pass in `self` as argument

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +529,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+      "extractAndRunWeightObserver can only be called on weight tensors");
+  Graph* g = observer->owningGraph();
+  std::vector<Node*> weight_subgraph;
+  remap_values_in_observer_subgraph_.clear();
```
This doesn't sound great, but I think we might be able to remove this if you move the code that clones a graph to a separate class

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/quantize_script.py:
```diff
@@ -78,5 +77,5 @@ def _quantize_script(model, qconfig_dict, run_fn=None, run_args=None, inplace=Fa
-def quantize_dynamic_script(model, qconfig_dict, sample_model_inputs, inplace=False, debug=False):
-    return _quantize_script(model, qconfig_dict, run_args=sample_model_inputs, inplace=inplace, debug=debug, is_dynamic=True)
+def quantize_dynamic_script(model, qconfig_dict, inplace=False, debug=False):
+    return _quantize_script(model, qconfig_dict, run_args=None, inplace=inplace, debug=debug, is_dynamic=True)
```
run_args=None can be removed now :)

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +547,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    Module& module,
+    Value* self,
+    Node* observer) {
+  Value* observed_weight = observer->output();
```
this is unused now

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -745,6 +873,49 @@ void propagateQuantizationOps(Block* block) {
+  // weight.
+  for (const auto& v : weight_values) {
+    ProcessWeightObserverSubgraph o;
+    o.extractAndRunWeightObserver(module, self, v->node());
```
I think it's better to cache the extracted graph so that we don't need to extract the same graph multiple times

jerryzh168 replied:
I feel it's better here to just use `v` as input, since we'll extract subgraph for this weight value.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +547,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  // If the graph was already visited, return list of relevant nodes directly.
+  // Multiple module instances can share the same graph code, so we don't need
+  // to re-run the extraction process.
+  if (graph_to_subgraph_nodes_.count(g)) {
```
the subgraph should not be indexed by graph, it should be indexed by value, we'll have one subgraph for each weight value

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]

supriyar commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +552,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  if (weight_to_subgraph_nodes.count(weight_value)) {
+    weight_subgraph = weight_to_subgraph_nodes[weight_value];
+  } else {
+    // Extract the subgraph nodes.
+    weight_subgraph.push_back(observer);
+    // Track values in the subgraph for which producer needs to be found.
+    std::vector<Value*> curr_node_inputs;
+    for (auto v : observer->inputs()) {
+      curr_node_inputs.push_back(v);
+    }
```
actually, 625-628 can be removed now. I forgot to delete it.

jerryzh168 replied:
it should probably be `findSubgraph(Value* v, std::vector<Node*>& weight_subgraph)`

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +552,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    std::vector<Node*>& weight_subgraph) {
+  for (auto v : inputs) {
+    auto n = v->node();
+    if (!hitGraphInput(n->output(0))) {
```
also, after we hit the graph input, we'll need to check if the input is same as `self` and error out if it's not: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#468

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +552,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    weight_to_subgraph_nodes[weight_value] = weight_subgraph;
+  }
+
+  auto observer_subgraph = std::make_shared<Graph>();
+  auto build_observer_graph = [&](Function& func) {
+    buildObserverSubgraph(weight_subgraph, func.graph());
+  };
+  auto func = torch::make_unique<GraphFunction>(
+      "observer_subgraph", observer_subgraph, build_observer_graph);
```
I think we can extract the building subgraph part to a separate class, and cache the graph function we build.

supriyar replied:
I thought about it before but couldn't find a way to do it since this part depends on the build function for the subgraph. Functionality wise it shouldn't make a difference imo. Lmk if you have a different view.

jerryzh168 replied:
not sure I understand why this is not possible, I think if we can extract this as a separate class, or I feel we maybe able to extract this as a function, we can put this in jit, other people might be able to use it as well.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +552,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  LintGraph(dest_graph);
+  // Add last node output value as subgraph output.
+  dest_graph->registerOutput(
+      updateInputValueInGraph(weight_subgraph.back()->output(0), dest_graph));
```
we can make this more general: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/ir/ir.cpp.html#661

jerryzh168 replied:
actually, is it possible to clone the graph with: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/ir/ir.cpp.html#_ZN5torch3jit5Graph11createCloneEPNS0_4NodeERKSt8functionIFPNS0_5ValueES6_EEb

supriyar replied:
@jerryzh168 I do use `createClone` please see https://github.com/pytorch/pytorch/pull/39412/files#diff-901c179fa606354f6bc42223debab96cR574

jerryzh168 replied:
I see. I looked at the code base and looks like the patterns are a little bit different: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/canonicalize.cpp.html#_ZN5torch3jit12CanonicalizeERKSt10shared_ptrINS0_5GraphEEb, could you follow the same pattern as the existing code?

jerryzh168 replied:
Also, should we check that the graph only has one input(%self)?

supriyar replied:
As long as the other input doesn't contribute to the weight observer we should be fine. I will add an assert for that in `findSubgraph`

jerryzh168 replied:
> As long as the other input doesn't contribute to the weight observer we should be fine. I will add an assert for that in `findSubgraph`

to clarify we should only accept the graph with one input since when we run the model we pass in a single input, maybe we should also check the input is %self as well, that is, make sure the type of the input matches the module.

supriyar replied:
This case should be fine imo as long as the extra inputs don't produce the weight, they wont be in the new graph.
Original graph 
```
graph(%self : __torch__.___torch_mangle_129.DynamicModel,
      %x.1 : Tensor,
      %x.2: Tensor):
...
%6 : Tensor = prim::GetAttr[name="weight"](%self)
..(%x.1 and %x.2 are used here but don't affect weight.
```
New graph
```
graph(%self : __torch__.___torch_mangle_129.DynamicModel):
...
%6 : Tensor = prim::GetAttr[name="weight"](%self)
(code to run observer forward)
```

jerryzh168 replied:
We have different graphs in mind, I'm talking about the extracted subgraph.

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -362,6 +362,34 @@ void checkCalculateQParamsResult(const IValue& qparams) {
+
+  // Map from values in original graph to corresponding values in the extracted
+  // subgraph.
+  std::unordered_map<Value*, Value*> remap_values_in_observer_subgraph_;
```
is it possible to remove this as a member variable?

supriyar replied:
Checked couple of other uses of graph clone functions in the codebase, they seem to have this as member variable.

jerryzh168 replied:
are you sure? I only saw local variable

jerryzh168 replied:
e.g. https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/codegen/cuda/graph_fuser.cpp.html#104,
https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/loop_unrolling.cpp.html#73,
https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/utils/subgraph_utils.cpp.html#158

supriyar replied:
Here is the example I was referring to. https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/bailout_graph.cpp.html#215. I think either way works. In this case since we are using different functions for each step we have a member variable. If we have one large function doing everything then it can be local.

jerryzh168 replied:
yeah, one large function sounds good

raghuramank100 replied:
I like the modularity of this class with the individual functions. If you are talking about only combining updateInputValueInGraph and cloneNodeInGraph, then I am fine either way.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    Module& module,
+    Value* self,
+    Value* weight_value,
+    std::unordered_map<Value*, std::vector<Node*>>& weight_to_subgraph_nodes) {
```
we don't need to pass in `weight_to_subgraph_nodes` anymore since it's a member variable

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  } else {
+    // Extract the subgraph nodes.
+    weight_subgraph.push_back(observer);
+    o.findSubgraph(self, weight_value, weight_subgraph);
```
do we need to put this function in ProcessWeightObserverSubgraph?

supriyar replied:
I think it is good to keep all related functions in the class. Serves as an entry point into building and running the graph for each value

jerryzh168 replied:
I feel it's cleaner to make `findSubgraph` a top level function or a member function for InsertQuantDequant, so that `ProcessWeightObserverSubgraph` can just be `SubgraphCloneHelper`

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  // Multiple module instances can share the same graph code, so we don't need
+  // to re-run the extraction process.
+  if (weight_to_subgraph_nodes.count(weight_value)) {
+    weight_subgraph = weight_to_subgraph_nodes[weight_value];
```
we can cache the function directly, so we don't need to rebuild the weight_subgraph every time

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    weight_subgraph = weight_to_subgraph_nodes[weight_value];
+  } else {
+    // Extract the subgraph nodes.
+    weight_subgraph.push_back(observer);
```
you can move this inside findSubgraph I think

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  const auto& inputs = input_val->node()->inputs().vec();
+  for (auto v : inputs) {
+    auto n = v->node();
+    if (!hitGraphInput(n->output(0))) {
```
why not just check `v` here?

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    Value* self,
+    Value* weight_value,
+    std::unordered_map<Value*, std::vector<Node*>>& weight_to_subgraph_nodes) {
+  Node* observer = weight_value->node();
+  Graph* g = observer->owningGraph();
```
please make sure to remove unused variables

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
   GRAPH_DUMP("After remove observers :", g);
 }
 
+Value* ProcessWeightObserverSubgraph::updateInputValueInGraph(
```
this can be a function local to `cloneNodeInGraph`

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    Value* self,
+    Value* input_val,
+    std::vector<Node*>& weight_subgraph) {
+  const auto& inputs = input_val->node()->inputs().vec();
```
I think you can push the node in the beginning

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +556,113 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  Node* observer = weight_value->node();
+  Graph* g = observer->owningGraph();
+  std::vector<Node*> weight_subgraph;
+  ProcessWeightObserverSubgraph o;
```
I feel this looks weird...if we refactor, we can remove this to later point

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2348,10 +2339,81 @@ def forward(self, x):
+
+            # Explicitly call forward on model before convert
+            m = prepare_dynamic_script(model, qconfig_dict)
+            m(data)
```
Why is this needed?

raghuramank100 replied:
Also, in this test how are we verifying the final graph? We need to check if dynamic quantization occurs as expected.  We will need to clone weights and quantize only one of them  (i.e only the weights used by the module that has a qconfig_dict)

supriyar replied:
This is how I make sure the numerics match between the two approaches, the `quantize_dynamic_script` call above doesn't take in the input data.

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2348,10 +2339,81 @@ def forward(self, x):
+                self.res2 = Res()
+
+            def forward(self, x):
+                x = self.res1(x, True)
```
For this test to be more useful, please test with dynamic inputs to forward. How about 
``` python 
def forward(self, x):
  cond=  torch.norm(x) > 10
  x = self.res1(x, cond)

supriyar replied:
There seems to be an issue with this in general (not related to my PR). Will debug later.

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2348,10 +2339,81 @@ def forward(self, x):
+        m2 = prepare_dynamic_script(model, qconfig_dict)
+        m2 = convert_dynamic_script(m2)
+        out_ref = m2(data)
+        self.assertEqual(out_graph, out_ref)
```
We need to explicitly check the graph of the dynamic quantized model, checking for consistency between prepare->convert and quantize is not sufficient as it will not catch errors made in both paths.

supriyar replied:
We are checking numerics here to make sure both ways work. The graph will not change since the extra weight subgraphs aren't added to the original graph.

raghuramank100 replied:
This is a general issue, but can be addressed in a later PR to check if we handle shared weights correctly

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +551,114 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    std::shared_ptr<Graph>& g,
+    std::unordered_map<Value*, Value*>& remap_old_to_new) {
+  auto* block = g->block();
+  auto value_fn = [&](Value* v) {
```
you can make `value_fn` `updateInputValueInGraph` I think

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +551,114 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+  weight_subgraph.push_back(node);
+  const auto& inputs = node->inputs().vec();
+  for (auto v : inputs) {
+    auto n = v->node();
```
you can remove `n` here

jerryzh168 reviewed (approved):
LGTM, thanks for addressing all the comments

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -511,6 +551,106 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    SubGraphCloneHelper o;
+    std::unique_ptr<GraphFunction> func =
+        o.buildGraphFromNodes(weight_subgraph, "observer_subgraph");
+    weight_to_graph_fn[weight_value] = std::move(func);
```
one final nit: `weight_to_graph_fn` -> `weight_to_graph_fn_`

supriyar committed:
Update on "[quant][graphmode] Dynamic Quant : Do not depend on input shapes"


Summary:
This PR introduces changes to enable running the weight observer standalone in the graph
It extracts the nodes from the graph that correspond to the observed weight value and adds all the related nodes to a new subgraph
The subgraph is then executed using GraphFunction

Test Plan:
python test/test_quantization.py TestGraphMostPostTrainingStatic
python test/test_quantization.py TestQuantizeDynamicScript

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21872940](https://our.internmc.facebook.com/intern/diff/D21872940)

[ghstack-poisoned]
